### PR TITLE
Fix #1619: Reduce collision rate in NameTable

### LINKED ISSUE #1619: Port NameTable collision rate reducing change from Scalac
https://github.com/scala/scala/pull/5474

I originally though we're not affected, but I've misunderstood the issue. We are affected.

### PR DESCRIPTION
**Pull request by jvican**
The fix is consistent with Scalac: https://github.com/scala/scala/pull/5474/files.

### CODE REVIEW

File: src/dotty/tools/dotc/core/Names.scala [hunk 1]
```diff
@@ -221,13 +221,15 @@ object Names {
...
-    if (len > 0)
-      (len * (41 * 41 * 41) +
-        cs(offset) * (41 * 41) +
-        cs(offset + len - 1) * 41 +
-        cs(offset + (len >> 1)))
-    else 0
+  private def hashValue(cs: Array[Char], offset: Int, len: Int): Int = {
+    var i = offset
+    var hash = 0
+    while (i < len) {
```
Comment (smarter) [inline src/dotty/tools/dotc/core/Names.scala]: I think you want `offset + len`

Comment (jvican) [inline src/dotty/tools/dotc/core/Names.scala]: Don't think so: https://github.com/scala/scala/pull/5474/files#diff-dd6d9f33ec7fb0a0b04f787bdb442bbdR47. We're already offsetting in `i`.

**Comment by jvican** [comment]
This change is useless until Dotty uses 2.12.1. Will reopen when 2.12.1 is out and you decide to bump up the Scala version.

File: src/dotty/tools/dotc/core/Names.scala (hunk 1 above)
Comment (smarter) [inline src/dotty/tools/dotc/core/Names.scala]: Original loop in the code you linked:
```scala
var i = 0
while (i < len) {
  h = ... cs(i + offset)
}
```
If you change `i` to start at `offset` then you also need to replace `len` by `len + offset`

**Comment by smarter** [comment]
> This change is useless until Dotty uses 2.12.1. Will reopen when 2.12.1 is out and you decide to bump up the Scala version.


Why?

File: src/dotty/tools/dotc/core/Names.scala (hunk 1 above)
Comment (jvican) [inline src/dotty/tools/dotc/core/Names.scala]: You're right, thanks!

**Comment by jvican** [comment]
To the best of my knowledge, tests were failing because Dotty depends on scalac when bootstrapping. If we have a new hashValue for the NameTable and the current Scala still uses the old one, then symbolic references change and are not available for Dotty.

**Comment by smarter** [comment]
No, these should be completely independent.

**Comment by jvican** [comment]
Errors:
```
[info] Running dotc.build 
------------  Building dotty  ------------
java.lang.Exception
	at dotty.tools.dotc.core.Denotations$MissingRef.<init>(Denotations.scala:1038)
	at dotty.tools.dotc.core.Denotations$DenotationsBase$class.recur$1(Denotations.scala:1193)
	at dotty.tools.dotc.core.Denotations$DenotationsBase$class.staticRef(Denotations.scala:1198)
	at dotty.tools.dotc.core.Contexts$ContextBase.staticRef(Contexts.scala:527)
	at dotty.tools.dotc.core.Symbols$class.requiredPackage(Symbols.scala:341)
	at dotty.tools.dotc.core.Contexts$Context.requiredPackage(Contexts.scala:57)
	at dotty.tools.dotc.core.Definitions.ScalaPackageVal$lzycompute(Definitions.scala:133)
	at dotty.tools.dotc.core.Definitions.ScalaPackageVal(Definitions.scala:133)
	at dotty.tools.dotc.core.Definitions.ScalaPackageClass$lzycompute(Definitions.scala:135)
	at dotty.tools.dotc.core.Definitions.ScalaPackageClass(Definitions.scala:135)
	at dotty.tools.dotc.core.Definitions.AnyClass$lzycompute(Definitions.scala:175)
	at dotty.tools.dotc.core.Definitions.AnyClass(Definitions.scala:175)
	at dotty.tools.dotc.core.Definitions.syntheticScalaClasses$lzycompute(Definitions.scala:772)
	at dotty.tools.dotc.core.Definitions.syntheticScalaClasses(Definitions.scala:771)
	at dotty.tools.dotc.core.Definitions.syntheticCoreClasses$lzycompute(Definitions.scala:782)
	at dotty.tools.dotc.core.Definitions.syntheticCoreClasses(Definitions.scala:782)
	at dotty.tools.dotc.core.Definitions.init(Definitions.scala:798)
	at dotty.tools.dotc.core.Contexts$ContextBase.initialize(Contexts.scala:570)
	at dotty.tools.dotc.Compiler.rootContext(Compiler.scala:118)
	at dotty.tools.dotc.Compiler.newRun(Compiler.scala:156)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:25)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at test.CompilerTest.compileArgs(CompilerTest.scala:250)
	at test.CompilerTest.compileDir(CompilerTest.scala:145)
	at dotc.tests.dotty(tests.scala:180)
	at dotc.build$.main(build.scala:26)
	at dotc.build.main(build.scala)
creating stub for scala, owner = module class <root>#2, file = null
decls = module xsbti#3
  module com#5
  module scala#7
  module dotty#9
  module org#11
  module scalaShadowing#13
  module strawman#15
  module sun#17
  module java#19
  module javax#21
  module jdk#23
  module bench#25
  module bin#27
  module bridge#29
  module examples#31
  module interfaces#33
  module out#35
  module project#37
  module sandbox#39
  module scripts#41
  module src#43
  module test#45
  module tests#47
  module scala-scala#49
  module scala-library#51
  module Test-dotty.obj#53
  module pos#55
  module linker-tests#57
  module target#59
  module dottydoc#61
  module docs#63

bad symbolic reference. A signature
refers to scala$ in package <root> which is not available.
It may be completely missing from the current classpath, or the version on
the classpath might be incompatible with the version used when compiling the signature.
Exception in thread "main" java.lang.UnsupportedOperationException: class dotty.tools.dotc.core.Scopes$EmptyScope$.openForMutations
	at dotty.tools.dotc.util.DotClass.unsupported(DotClass.scala:10)
	at dotty.tools.dotc.core.Scopes$Scope.openForMutations(Scopes.scala:142)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.enter(SymDenotations.scala:1514)
	at dotty.tools.dotc.core.Symbols$Symbol.entered(Symbols.scala:430)
	at dotty.tools.dotc.core.Definitions.newCompleteClassSymbol(Definitions.scala:51)
	at dotty.tools.dotc.core.Definitions.AnyClass$lzycompute(Definitions.scala:175)
	at dotty.tools.dotc.core.Definitions.AnyClass(Definitions.scala:175)
	at dotty.tools.dotc.core.Definitions.syntheticScalaClasses$lzycompute(Definitions.scala:772)
	at dotty.tools.dotc.core.Definitions.syntheticScalaClasses(Definitions.scala:771)
	at dotty.tools.dotc.core.Definitions.syntheticCoreClasses$lzycompute(Definitions.scala:782)
	at dotty.tools.dotc.core.Definitions.syntheticCoreClasses(Definitions.scala:782)
	at dotty.tools.dotc.core.Definitions.init(Definitions.scala:798)
	at dotty.tools.dotc.core.Contexts$ContextBase.initialize(Contexts.scala:570)
	at dotty.tools.dotc.Compiler.rootContext(Compiler.scala:118)
	at dotty.tools.dotc.Compiler.newRun(Compiler.scala:156)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:25)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at test.CompilerTest.compileArgs(CompilerTest.scala:250)
	at test.CompilerTest.compileDir(CompilerTest.scala:145)
	at dotc.tests.dotty(tests.scala:180)
	at dotc.build$.main(build.scala:26)
	at dotc.build.main(build.scala)
java.lang.RuntimeException: Nonzero exit code returned from runner: 1
	at scala.sys.package$.error(package.scala:27)
```
  
What could be this error then?

**Comment by jvican** [comment]
And this is the CI: https://scala-ci.typesafe.com/job/dotty-master-validate-partest-bootstrapped/1689/console.

**Comment by smarter** [comment]
Does that still happen after fixing the bug I mentioned above?

**Comment by jvican** [comment]
No, was on my way of testing it. Thanks for the report, reopening in another PR.

**Pull request closed without merging**